### PR TITLE
Add retry engine

### LINKED ISSUE #178: Lots of unreleased code, any plans for a deployment to crates soon?
Hey, I noticed a lot of new commits coming in and I was wondering when they would be pushed to crates.io. Specifically looking to pick up the change where the `card` field was added to `CreatePaymentMethod`.

### PR DESCRIPTION
**Pull request by arlyon**
This PR adds the ability to have the client automatically apply retries by applying a `RequestStrategy`. Closes #178 

- [x] refactor common client code into single struct
- [x] add idempotency strategy
- [x] add retries strategy 
- [x] port async-std client
- [x] port tokio and tokio-blocking clients 
- [x] restore set_stripe_account api
- [x] determine time-out function
- [x] fill out documentation
- [ ] add examples
- [x] clone requests correctly

There are 4 strategies:

## Once

This is the basic usage (and the default). This strategy will attempt to make a request, and immediately return control to the caller when it fails.

## Idempotent

This is one step up from `Once`. Using this strategy will use a idempotency key of your choice, and make requests with it. Failures will continue to return control to the caller, however you may retry the request at will.

## Retry (not recommended)

This strategy will make the same request up to `n` times, immediately one-after-the-other. This strategy is not recommended for production use as it is bad manners. Regardless, it may be useful for testing.

## Exponential Back-off

This strategy inherits all the semantics of the `Retry` strategy, but with an increasing wait between requests.

> 🚨 Note: in both the `Retry` and `ExponentialBackoff` strategies `n` is the maximum limit. If the stripe API explicitly communicates to the client that a retry is not useful (via the `Stripe-Should-Retry` header), the request will be aborted without reaching the limit.

# Usage

```rust
// create a client, and use the ExponentialBackoff strategy for all subsequent requests
let client = Client::new("sk_1234").with_strategy(RequestStrategy::ExponentialBackoff(5));

// we need manual control, let's make a new client and use the idempotent strategy
let idempotent = client.clone().with_strategy(RequestStrategy::Idempotent("my_token"));
```

### CODE REVIEW

**Comment by arlyon** [comment]
@FL33TW00D ^

File: src/client/base/tokio.rs [hunk 1]
```diff
@@ -12,8 +12,6 @@ use crate::error::{ErrorResponse, StripeError};
 use crate::params::{AppInfo, Headers};
 use crate::resources::ApiVersion;
 
-static USER_AGENT: &str = concat!("Stripe/v3 RustBindings/", env!("CARGO_PKG_VERSION"));
```
Comment (FL33TW00D) [inline src/client/base/tokio.rs:15]: Removed this intentionally?

**Review by FL33TW00D — COMMENTED** [review]
Looks really clean, few things failing but apart from that.

**Comment by arlyon** [comment]
Missing some final issues around cloning requests. Basically cloning an http-types::Request truncates the body, so you need to re-add it. Not a tough fix but it'll have to wait for the moment!

**Pull request merged**